src/AuditTrailModule/javasource/test_crm/tests/expected/ExpectedReferenceLog.java

```diff
@@ -0,0 +1,56 @@
+package test_crm.tests.expected;
+
+import static org.junit.Assert.*;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import com.mendix.systemwideinterfaces.core.IContext;
+import com.mendix.systemwideinterfaces.core.IMendixObject;
+import com.mendix.systemwideinterfaces.core.IMendixObjectMember;
+
+import audittrail.proxies.ReferenceLog;
+import audittrail.proxies.ReferenceLogLine;
+import audittrail.proxies.TypeOfReferenceLog;
+import test_crm.tests.actual.ActualReferenceLog;
+
+public class ExpectedReferenceLog {
+	private final TypeOfReferenceLog operation;
+	private final Long referenceLogId;
+	private final Map<String, String> memberValues;
+
+	public ExpectedReferenceLog(final IContext context, final IMendixObject referencedObject, final TypeOfReferenceLog operation) {
+		this.referenceLogId = referencedObject.getId().toLong();
+		this.operation = operation;
+
+		this.memberValues = new HashMap<String, String>();
+		for (final IMendixObjectMember<?>  obj : referencedObject.getMembers(context).values()) {
+			memberValues.put(obj.getName(), obj.getValue(context).toString());
+		}
+	}
+
+	public void verify(final ActualReferenceLog actualLogLine) {
+		final ReferenceLog reference = actualLogLine.getReferenceLog();
+
+		assertEquals("ID mismatch when comparing reference logs.", this.referenceLogId.toString(), reference.getAttributeID());
+		assertEquals("Incorrect reference log operation for reference to object of ID " + reference.getAttributeID(), this.operation, reference.getOperation());
+
+		// Now verify loglines:
+		assertEquals("Incorrect number of loglines for reference.", this.memberValues.size(), actualLogLine.getReferenceLogLines().size());
+		final Set<String> visitedLogLines = new HashSet<String>();
+		for (final ReferenceLogLine logLine : actualLogLine.getReferenceLogLines()) {
+			final String memberName = logLine.getMember();
+			// Have we seen this one before?
+			assertFalse("Repeated log line for member " + memberName, visitedLogLines.contains(memberName));
```

Comment:
Same as before: it's better to compare lists of names with Hamcrest

Reply:
This and the other hamcrest comments should be handled by now.